app/utils/useContextMenu.tsx

```diff
@@ -0,0 +1,137 @@
+  id?: string
+  label: string
+  shortcut?: string
+  enabled?: boolean
+  checked?: boolean
+  children?: MenuListEntry[]
+  action?: () => AsyncableVoid
+}
+
+export type MenuListEntry = MenuItem | typeof SEPARATOR
+
+export interface ContextMenuConfig {
+  items: MenuListEntry[]
+}
+
+// used to separate path components in the action map key
+const DELIM = "\x1F" // 0x1F is the ASCII record separator character
+const makeKey = (path: readonly string[]) => path.join(DELIM)
+
+type NativeMenuItem = {
+  label: string
+  shortcut?: string
+  enabled?: boolean
+  checked?: boolean
+  children?: NativeMenuListEntry[]
+}
+
+type NativeMenuListEntry = NativeMenuItem | typeof SEPARATOR
+
+export function useContextMenu(config: ContextMenuConfig) {
+  // We keep the latest actionMap in a ref so onPress doesn't resubscribe.
+  const actionsRef = useRef<Map<string, () => AsyncableVoid>>(new Map())
+
+  // Build both the sanitized items (for native) and the action map (for lookup) in one pass.
+  const { nativeItems, actionMap } = useMemo(() => {
+    const nextMap = new Map<string, () => AsyncableVoid>()
+
+    const build = (entries: MenuListEntry[], parent: string[] = []): NativeMenuListEntry[] => {
+      const out: NativeMenuListEntry[] = []
+
+      const pushEntry = (e: NativeMenuListEntry) => {
+        // prevent consecutive separators
+        if (e === SEPARATOR && out[out.length - 1] === SEPARATOR) return
+        out.push(e)
+      }
+
+      for (const entry of entries) {
+        if (entry === SEPARATOR) {
+          pushEntry(SEPARATOR)
+          continue
+        }
+
+        const item = entry as MenuItem
+        const { label, shortcut, enabled, checked, children, action } = item
+
+        // Prefer explicit ids in the path; fall back to label if not provided
+        const nodeId = item.id ?? label
+        const path = [...parent, nodeId]
+
+        if (typeof action === "function") {
+          const key = makeKey(path)
+          if (__DEV__ && nextMap.has(key)) {
+            // Useful when two sibling items share id/label
+            console.warn(`Duplicate context-menu path: ${path.join(" > ")}`)
+          }
+          nextMap.set(key, action)
+        }
+
+        pushEntry({
+          label, // native UI shows labels, ids are just for our lookup
+          shortcut,
+          enabled,
+          checked,
+          children: Array.isArray(children) ? build(children, path) : undefined,
+        })
+      }
+
+      // trim trailing separator
+      while (out.length && out[out.length - 1] === SEPARATOR) out.pop()
+      // trim leading separator
+      while (out.length && out[0] === SEPARATOR) out.shift()
+
+      return out
+    }
+
+    const built = build(config.items)
+    return { nativeItems: built, actionMap: nextMap }
+  }, [config.items])
+
+  // Keep the latest map in a ref without changing onPress identity.
+  useEffect(() => {
+    actionsRef.current = actionMap
+  }, [actionMap])
```

Comment:
I could be wrong, but it seems like we could remove the `useEffect` here and just do this:

```suggestion
  // Keep the latest map in a ref without changing onPress identity.
  actionsRef.current = actionMap
```

I don't think the useEffect is doing anything useful, here.

Again, I could be wrong. useEffect is not the easiest to grok

Reply:
Ahhh ... wild, I didn't know that.

Reply:
I _think_ we would lose our referential stability from that which would lead to more sub/unsubs. Probably not a huge deal but if the useMemo changes I think it would return a new reference to onPress which would cause the useEffect that handles subscribe to see an onPress change and rerun the subscribe code.